File: Core/src/main/java/dev/lynith/core/Logger.java
package dev.lynith.core;

public class Logger {

    private final String name;

    public Logger(String name) {
        this.name = name;
    }

    public Logger(String... name) {
        this(String.join(".", name));
    }

    public Logger(Class<?> clazz) {
        this(clazz.getSimpleName());
    }

    public Logger(Object obj) {
        this(obj.getClass());
    }

    public Logger() {
        this("unknown");
    }

    public void log(Object message) {
        System.out.println(getString(message.toString()));
    }

    public void log(Object message, Object... args) {
        System.out.printf((getString(message.toString())) + "%n", args);
    }

    public void error(Object message) {
        System.err.println(getString(message.toString()));
    }

    public void error(Object message, Object... args) {
        System.err.printf((getString(message.toString())) + "%n", args);
    }

    private String getString(String message) {
        return "[" + name + "] " + message;
    }

}

File: Core/src/main/kotlin/dev/lynith/core/Logger.kt
package dev.lynith.core

class Logger @JvmOverloads constructor(private val name: String = "unknown") {
    constructor(vararg name: String?) : this(java.lang.String.join(".", *name))
    constructor(clazz: Class<*>) : this(clazz.getSimpleName())
    constructor(obj: Any) : this(obj.javaClass)

    fun log(message: Any) {
        println(getString(message.toString()))
    }

    fun log(message: Any, vararg args: Any) {
        println(getString(format(message.toString(), *args)))
    }

    fun error(message: Any) {
        System.err.println(getString(message.toString()))
    }

    fun error(message: Any, vararg args: Any) {
        System.err.println(getString(format(message.toString(), *args)))
    }

    private fun format(message: String, vararg args: Any): String {
        var message = message
        for (arg in args) {
            val argString = arg.toString()
            message = message.replaceFirst("\\{\\}".toRegex(), argString)
        }
        return message
    }

    private fun getString(message: String?): String {
        return "[$name] $message"
    }
}
